File: scripts_archive/node/scripts/seed_appellation_aliases.mjs
#!/usr/bin/env node
/**
 * seed_appellation_aliases.mjs — Populate appellation_aliases from primary sources + mechanical generation
 *
 * Sources:
 *   1. INAO OpenDataSoft API — French AOC/IGP product variants (color, cru, style)
 *   2. Eurac PDO_EU_cat.csv — EU PDO wine product categories
 *   3. Tier 1 mechanical generation — color suffixes, accent-stripped, designation types
 *   4. Tier 2 known translations — English ↔ local name pairs
 *
 * Usage: node scripts/seed_appellation_aliases.mjs [--dry-run]
 */

import { readFileSync, writeFileSync } from 'fs';
import { createClient } from '@supabase/supabase-js';

// ── Load .env ───────────────────────────────────────────────
const envPath = new URL('../.env', import.meta.url).pathname.replace(/^\/([A-Z]:)/, '$1');
const envContent = readFileSync(envPath, 'utf-8');
for (const line of envContent.split('\n')) {
  const trimmed = line.replace(/\r/g, '').trim();
  if (!trimmed || trimmed.startsWith('#')) continue;
  const eqIdx = trimmed.indexOf('=');
  if (eqIdx === -1) continue;
  const key = trimmed.slice(0, eqIdx);
  const val = trimmed.slice(eqIdx + 1);
  if (!process.env[key]) process.env[key] = val;
}

const SUPABASE_URL = process.env.SUPABASE_URL;
const SUPABASE_KEY = process.env.SUPABASE_SERVICE_ROLE || process.env.SUPABASE_ANON_KEY;
const sb = createClient(SUPABASE_URL, SUPABASE_KEY);
const DRY_RUN = process.argv.includes('--dry-run');

// ── Helpers ─────────────────────────────────────────────────
function normalize(s) {
  return s.normalize('NFD').replace(/[\u0300-\u036f]/g, '').toLowerCase()
    .replace(/[^a-z0-9 ]/g, ' ').replace(/\s+/g, ' ').trim();
}

async function fetchAll(table, columns = '*', filter = {}, batchSize = 1000) {
  const all = [];
  let offset = 0;
  while (true) {
    let query = sb.from(table).select(columns).range(offset, offset + batchSize - 1);
    for (const [k, v] of Object.entries(filter)) query = query.eq(k, v);
    const { data, error } = await query;
    if (error) throw new Error(`fetchAll ${table}: ${error.message}`);
    all.push(...data);
    if (data.length < batchSize) break;
    offset += batchSize;
  }
  return all;
}

// ── Color suffix maps per country ───────────────────────────
const COLOR_SUFFIXES = {
  // French
  FR: ['rouge', 'blanc', 'rosé', 'clairet'],
  // Italian
  IT: ['rosso', 'bianco', 'rosato'],
  // Spanish
  ES: ['tinto', 'blanco', 'rosado'],
  // Portuguese
  PT: ['tinto', 'branco', 'rosado', 'rosé'],
  // German — less common but exists
  DE: ['rot', 'weiß', 'weiss', 'rosé'],
  // Austrian
  AT: ['rot', 'weiß', 'weiss', 'rosé'],
  // Generic English (for New World)
  US: ['red', 'white', 'rosé', 'rose'],
  AU: ['red', 'white', 'rosé', 'rose'],
  NZ: ['red', 'white', 'rosé', 'rose'],
  ZA: ['red', 'white', 'rosé', 'rose'],
  CL: ['tinto', 'blanco', 'rosado'],
  AR: ['tinto', 'blanco', 'rosado'],
};

// ── Known English ↔ local translations ──────────────────────
// Format: [english_variant, local_name, iso_code]
// These are appellation-level names that appear differently in English vs local language
const TRANSLATIONS = [
  // French
  ['Burgundy', 'Bourgogne', 'FR'],
  ['Rhone Valley', 'Vallée du Rhône', 'FR'],
  ['Rhone', 'Rhône', 'FR'],
  ['Cotes du Rhone', 'Côtes du Rhône', 'FR'],
  ['Cotes du Rhone Villages', 'Côtes du Rhône Villages', 'FR'],
  ['Beaujolais Villages', 'Beaujolais-Villages', 'FR'],
  ['Cote de Beaune', 'Côte de Beaune', 'FR'],
  ['Cote de Nuits', 'Côte de Nuits', 'FR'],
  ['Cotes de Provence', 'Côtes de Provence', 'FR'],
  ['Cotes du Roussillon', 'Côtes du Roussillon', 'FR'],
  ['Coteaux du Layon', 'Coteaux du Layon', 'FR'],
  ['Saint-Emilion', 'Saint-Émilion', 'FR'],
  ['Pouilly-Fume', 'Pouilly-Fumé', 'FR'],
  ['Pouilly-Fuisse', 'Pouilly-Fuissé', 'FR'],
  ['Chateauneuf-du-Pape', 'Châteauneuf-du-Pape', 'FR'],
  ['Cote Rotie', 'Côte Rôtie', 'FR'],
  ['Cote-Rotie', 'Côte Rôtie', 'FR'],
  ['Gevrey-Chambertin', 'Gevrey-Chambertin', 'FR'], // same but accent variants exist
  ['Cremant d\'Alsace', 'Crémant d\'Alsace', 'FR'],
  ['Cremant de Bourgogne', 'Crémant de Bourgogne', 'FR'],
  ['Cremant de Loire', 'Crémant de Loire', 'FR'],
  ['Cremant de Limoux', 'Crémant de Limoux', 'FR'],
  ['Corbieres', 'Corbières', 'FR'],
  ['Fitou', 'Fitou', 'FR'],
  ['Medoc', 'Médoc', 'FR'],
  ['Haut-Medoc', 'Haut-Médoc', 'FR'],
  ['Premieres Cotes de Bordeaux', 'Premières Côtes de Bordeaux', 'FR'],
  ['Entre-Deux-Mers', 'Entre-deux-Mers', 'FR'],
  ['Cotes de Bourg', 'Côtes de Bourg', 'FR'],
  ['Cotes de Blaye', 'Côtes de Blaye', 'FR'],
  // Italian
  ['Piedmont', 'Piemonte', 'IT'],
  ['Tuscany', 'Toscana', 'IT'],
  ['Brunello di Montalcino', 'Brunello di Montalcino', 'IT'],
  ['Vino Nobile di Montepulciano', 'Vino Nobile di Montepulciano', 'IT'],
  // Spanish
  ['Rioja', 'Rioja', 'ES'],
  ['Sherry', 'Jerez-Xérès-Sherry', 'ES'],
  ['Jerez', 'Jerez-Xérès-Sherry', 'ES'],
  ['Cava', 'Cava', 'ES'],
  ['Priorat', 'Priorat', 'ES'],
  ['Priorato', 'Priorat', 'ES'],
  // Portuguese
  ['Port', 'Porto', 'PT'],
  ['Oporto', 'Porto', 'PT'],
  ['Douro', 'Douro', 'PT'],
  ['Madeira', 'Madeira', 'PT'],
  ['Vinho Verde', 'Vinho Verde', 'PT'],
  ['Dao', 'Dão', 'PT'],
  // German
  ['Mosel', 'Mosel', 'DE'],
  ['Moselle', 'Mosel', 'DE'],
  ['Rhine', 'Rheingau', 'DE'],
  ['Pfalz', 'Pfalz', 'DE'],
  ['Palatinate', 'Pfalz', 'DE'],
  // Austrian
  ['Wachau', 'Wachau', 'AT'],
  // Hungarian
  ['Tokaj', 'Tokaji', 'HU'],
  ['Tokay', 'Tokaji', 'HU'],
  // Greek
  ['Santorini', 'Σαντορίνη', 'GR'],
  ['Naoussa', 'Νάουσα', 'GR'],
  ['Nemea', 'Νεμέα', 'GR'],
];

// ── Main ────────────────────────────────────────────────────
async function main() {
  console.log(`\n${'='.repeat(60)}`);
  console.log('  APPELLATION ALIASES — SEED FROM PRIMARY SOURCES');
  console.log(`  ${DRY_RUN ? '(DRY RUN)' : '(LIVE)'}`);
  console.log(`${'='.repeat(60)}\n`);

  // Load all appellations with country info
  const appellations = await fetchAll('appellations', 'id,name,designation_type,country_id');
  const countries = await fetchAll('countries', 'id,name,iso_code');
  const countryMap = new Map(countries.map(c => [c.id, c]));
  const isoToCountryId = new Map(countries.map(c => [c.iso_code, c.id]));

  // Build appellation lookup by normalized name
  const appellationByName = new Map();
  const appellationByNorm = new Map();
  for (const a of appellations) {
    appellationByName.set(a.name.toLowerCase(), a);
    appellationByNorm.set(normalize(a.name), a);
  }

  console.log(`Loaded ${appellations.length} appellations from ${countries.length} countries\n`);

  // Collect all aliases to insert
  const aliases = []; // { appellation_id, alias, alias_normalized, alias_type, source }
  const seen = new Set(); // dedupe by alias_normalized

  function addAlias(appellationId, alias, aliasType, source) {
    const norm = normalize(alias);
    // Skip if same as the canonical name
    const appellation = appellations.find(a => a.id === appellationId);
    if (appellation && normalize(appellation.name) === norm) return;
    // Skip duplicates
    if (seen.has(norm)) return;
    seen.add(norm);
    aliases.push({
      appellation_id: appellationId,
      alias,
      alias_normalized: norm,
      alias_type: aliasType,
      source,
    });
  }

  // ════════════════════════════════════════════════════════════
  // SOURCE 1: INAO OpenDataSoft API — French product variants
  // ════════════════════════════════════════════════════════════
  console.log('1. Fetching INAO product variants...');

  const INAO_API = 'https://public.opendatasoft.com/api/explore/v2.1/catalog/datasets/aires-et-produits-aocaop-et-igp/records';
  let inaoProducts = [];
  let inaoOffset = 0;
  const INAO_LIMIT = 100;

  // Filter to wine products only (AOC/IGP designation)
  const INAO_WHERE = encodeURIComponent("signe_fr LIKE 'AOC%' OR signe_fr LIKE 'IGP%'");

  while (true) {
    const url = `${INAO_API}?select=aire_geographique,produit,signe_fr,idproduit&group_by=aire_geographique,produit,signe_fr,idproduit&where=${INAO_WHERE}&limit=${INAO_LIMIT}&offset=${inaoOffset}`;
    const resp = await fetch(url);
    const data = await resp.json();
    if (!data.results || data.results.length === 0) break;
    inaoProducts.push(...data.results);
    inaoOffset += INAO_LIMIT;
    if (data.results.length < INAO_LIMIT) break;
    // Rate limit
    await new Promise(r => setTimeout(r, 200));
  }

  console.log(`  Fetched ${inaoProducts.length} INAO wine product entries`);

  // Match INAO products to our appellations
  let inaoMatched = 0;
  let inaoUnmatched = 0;
  const inaoMissedSet = new Set();

  for (const p of inaoProducts) {
    const aireNorm = normalize(p.aire_geographique);
    const produit = p.produit;
    const produitNorm = normalize(produit);

    // Skip if produit is same as aire_geographique (no variant)
    if (aireNorm === produitNorm) continue;

    // Find matching appellation by aire_geographique
    let appellation = appellationByNorm.get(aireNorm);

    // If no direct match, try matching just the base appellation name
    // (INAO has "Aloxe-Corton premier cru Clos des Maréchaudes" as aire_geographique
    //  but we might only have "Aloxe-Corton")
    if (!appellation) {
      // Try progressively shorter prefixes
      const words = p.aire_geographique.split(' ');
      for (let len = words.length - 1; len >= 1; len--) {
        const prefix = words.slice(0, len).join(' ');
        appellation = appellationByNorm.get(normalize(prefix));
        if (appellation) break;
      }
    }

    if (appellation) {
      // Determine alias type based on what's different
      let aliasType = 'synonym';
      const diff = produit.replace(p.aire_geographique, '').trim().toLowerCase();
      if (['rouge', 'blanc', 'rosé', 'clairet'].includes(diff)) aliasType = 'with_color';
      else if (diff.includes('vendanges tardives') || diff.includes('sélection de grains nobles')
        || diff.includes('vin jaune') || diff.includes('vin de paille')
        || diff.includes('mousseux') || diff.includes('primeur')
        || diff.includes('supérieur') || diff.includes('grand cru')
        || diff.includes('premier cru')) aliasType = 'with_designation';

      addAlias(appellation.id, produit, aliasType, 'inao-opendatasoft');
      inaoMatched++;
    } else {
      inaoUnmatched++;
      inaoMissedSet.add(p.aire_geographique);
    }
  }

  console.log(`  Matched: ${inaoMatched}, Unmatched: ${inaoUnmatched} (${inaoMissedSet.size} unique aires)`);
  if (inaoMissedSet.size > 0) {
    console.log(`  Sample unmatched: ${[...inaoMissedSet].slice(0, 10).join(', ')}`);
  }

  // ════════════════════════════════════════════════════════════
  // SOURCE 2: Tier 1 — Mechanical color suffix generation
  // ════════════════════════════════════════════════════════════
  console.log('\n2. Generating Tier 1 color suffix aliases...');

  let colorCount = 0;
  for (const a of appellations) {
    const country = countryMap.get(a.country_id);
    if (!country || !country.iso_code) continue;
    const suffixes = COLOR_SUFFIXES[country.iso_code];
    if (!suffixes) continue;

    for (const suffix of suffixes) {
      addAlias(a.id, `${a.name} ${suffix}`, 'with_color', 'mechanical-color-suffix');
      colorCount++;
    }
  }
  console.log(`  Generated ${colorCount} color suffix candidates (${aliases.length - colorCount} already existed from INAO)`);

  // ════════════════════════════════════════════════════════════
  // SOURCE 3: Tier 1 — Accent-stripped variants
  // ════════════════════════════════════════════════════════════
  console.log('\n3. Generating accent-stripped aliases...');

  let accentCount = 0;
  for (const a of appellations) {
    const stripped = a.name.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
    if (stripped !== a.name) {
      addAlias(a.id, stripped, 'synonym', 'mechanical-accent-strip');
      accentCount++;
    }
  }
  console.log(`  Generated ${accentCount} accent-stripped aliases`);

  // ════════════════════════════════════════════════════════════
  // SOURCE 4: Tier 1 — Designation type suffixes
  // ════════════════════════════════════════════════════════════
  console.log('\n4. Generating designation type suffix aliases...');

  let designationCount = 0;
  for (const a of appellations) {
    if (!a.designation_type) continue;
    // Skip if the name already contains the designation
    if (a.name.toLowerCase().includes(a.designation_type.toLowerCase())) continue;
    addAlias(a.id, `${a.name} ${a.designation_type}`, 'with_designation', 'mechanical-designation-suffix');
    designationCount++;
  }
  console.log(`  Generated ${designationCount} designation suffix aliases`);

  // ════════════════════════════════════════════════════════════
  // SOURCE 5: Tier 2 — Known translations
  // ════════════════════════════════════════════════════════════
  console.log('\n5. Adding known translation aliases...');

  let translationCount = 0;
  for (const [english, local, iso] of TRANSLATIONS) {
    // Find the canonical appellation — could be either the english or local name
    let appellation = appellationByName.get(local.toLowerCase())
      || appellationByName.get(english.toLowerCase());
    if (!appellation) {
      // Try normalized
      appellation = appellationByNorm.get(normalize(local))
        || appellationByNorm.get(normalize(english));
    }
    if (appellation) {
      // Add the variant that ISN'T the canonical name
      if (normalize(english) !== normalize(appellation.name)) {
        addAlias(appellation.id, english, 'synonym', 'known-translation');
        translationCount++;
      }
      if (normalize(local) !== normalize(appellation.name)) {
        addAlias(appellation.id, local, 'local_name', 'known-translation');
        translationCount++;
      }
    }
  }
  console.log(`  Added ${translationCount} translation aliases`);

  // ════════════════════════════════════════════════════════════
  // INSERT INTO DATABASE
  // ════════════════════════════════════════════════════════════
  console.log(`\n${'─'.repeat(60)}`);
  console.log(`Total aliases to insert: ${aliases.length}`);

  if (DRY_RUN) {
    console.log('\n(DRY RUN — not inserting)\n');
    // Show sample
    console.log('Sample aliases:');
    const samples = aliases.filter(a => a.source === 'inao-opendatasoft').slice(0, 10);
    for (const s of samples) {
      const app = appellations.find(a => a.id === s.appellation_id);
      console.log(`  "${s.alias}" → ${app?.name} [${s.alias_type}] (${s.source})`);
    }
    console.log('...');
    const colorSamples = aliases.filter(a => a.source === 'mechanical-color-suffix').slice(0, 5);
    for (const s of colorSamples) {
      const app = appellations.find(a => a.id === s.appellation_id);
      console.log(`  "${s.alias}" → ${app?.name} [${s.alias_type}] (${s.source})`);
    }

    // Stats by source
    const bySrc = {};
    for (const a of aliases) bySrc[a.source] = (bySrc[a.source] || 0) + 1;
    console.log('\nBy source:');
    for (const [src, count] of Object.entries(bySrc).sort((a, b) => b[1] - a[1])) {
      console.log(`  ${src}: ${count}`);
    }
    return;
  }

  // Batch insert
  const BATCH = 500;
  let inserted = 0;
  for (let i = 0; i < aliases.length; i += BATCH) {
    const batch = aliases.slice(i, i + BATCH);
    const { error } = await sb.from('appellation_aliases').upsert(batch, {
      onConflict: 'alias_normalized',
      ignoreDuplicates: true,
    });
    if (error) {
      console.error(`  Batch error at ${i}: ${error.message}`);
      // Try one by one for this batch
      for (const a of batch) {
        const { error: e2 } = await sb.from('appellation_aliases').upsert(a, {
          onConflict: 'alias_normalized',
          ignoreDuplicates: true,
        });
        if (!e2) inserted++;
      }
    } else {
      inserted += batch.length;
    }
  }

  console.log(`\nInserted ${inserted} aliases`);

  // Final stats
  const { data: finalCount } = await sb.from('appellation_aliases').select('source', { count: 'exact', head: false });
  const bySrc = {};
  for (const row of finalCount || []) bySrc[row.source] = (bySrc[row.source] || 0) + 1;
  console.log('\nFinal counts by source:');
  for (const [src, count] of Object.entries(bySrc).sort((a, b) => b[1] - a[1])) {
    console.log(`  ${src}: ${count}`);
  }
}

main().catch(console.error);

File: pipeline/reference/seed_appellation_aliases.py
"""
Populate appellation_aliases from primary sources + mechanical generation.

Sources:
  1. INAO OpenDataSoft API -- French AOC/IGP product variants (color, cru, style)
  2. Tier 1 mechanical generation -- color suffixes, accent-stripped, designation types
  3. Tier 2 known translations -- English <-> local name pairs

Usage:
    python -m pipeline.reference.seed_appellation_aliases [--dry-run]
"""

import argparse
import sys
import time
from pathlib import Path

# Allow running as module from repo root
sys.path.insert(0, str(Path(__file__).resolve().parents[2]))

import httpx

from pipeline.lib.db import get_supabase, fetch_all
from pipeline.lib.normalize import normalize


# ---- Color suffix maps per country ----
COLOR_SUFFIXES = {
    "FR": ["rouge", "blanc", "rosé", "clairet"],
    "IT": ["rosso", "bianco", "rosato"],
    "ES": ["tinto", "blanco", "rosado"],
    "PT": ["tinto", "branco", "rosado", "rosé"],
    "DE": ["rot", "weiß", "weiss", "rosé"],
    "AT": ["rot", "weiß", "weiss", "rosé"],
    "US": ["red", "white", "rosé", "rose"],
    "AU": ["red", "white", "rosé", "rose"],
    "NZ": ["red", "white", "rosé", "rose"],
    "ZA": ["red", "white", "rosé", "rose"],
    "CL": ["tinto", "blanco", "rosado"],
    "AR": ["tinto", "blanco", "rosado"],
}

# ---- Known English <-> local translations ----
# Format: (english_variant, local_name, iso_code)
TRANSLATIONS = [
    # French
    ("Burgundy", "Bourgogne", "FR"),
    ("Rhone Valley", "Vallée du Rhône", "FR"),
    ("Rhone", "Rhône", "FR"),
    ("Cotes du Rhone", "Côtes du Rhône", "FR"),
    ("Cotes du Rhone Villages", "Côtes du Rhône Villages", "FR"),
    ("Beaujolais Villages", "Beaujolais-Villages", "FR"),
    ("Cote de Beaune", "Côte de Beaune", "FR"),
    ("Cote de Nuits", "Côte de Nuits", "FR"),
    ("Cotes de Provence", "Côtes de Provence", "FR"),
    ("Cotes du Roussillon", "Côtes du Roussillon", "FR"),
    ("Coteaux du Layon", "Coteaux du Layon", "FR"),
    ("Saint-Emilion", "Saint-Émilion", "FR"),
    ("Pouilly-Fume", "Pouilly-Fumé", "FR"),
    ("Pouilly-Fuisse", "Pouilly-Fuissé", "FR"),
    ("Chateauneuf-du-Pape", "Châteauneuf-du-Pape", "FR"),
    ("Cote Rotie", "Côte Rôtie", "FR"),
    ("Cote-Rotie", "Côte Rôtie", "FR"),
    ("Gevrey-Chambertin", "Gevrey-Chambertin", "FR"),
    ("Cremant d'Alsace", "Crémant d'Alsace", "FR"),
    ("Cremant de Bourgogne", "Crémant de Bourgogne", "FR"),
    ("Cremant de Loire", "Crémant de Loire", "FR"),
    ("Cremant de Limoux", "Crémant de Limoux", "FR"),
    ("Corbieres", "Corbières", "FR"),
    ("Fitou", "Fitou", "FR"),
    ("Medoc", "Médoc", "FR"),
    ("Haut-Medoc", "Haut-Médoc", "FR"),
    ("Premieres Cotes de Bordeaux", "Premières Côtes de Bordeaux", "FR"),
    ("Entre-Deux-Mers", "Entre-deux-Mers", "FR"),
    ("Cotes de Bourg", "Côtes de Bourg", "FR"),
    ("Cotes de Blaye", "Côtes de Blaye", "FR"),
    # Italian
    ("Piedmont", "Piemonte", "IT"),
    ("Tuscany", "Toscana", "IT"),
    ("Brunello di Montalcino", "Brunello di Montalcino", "IT"),
    ("Vino Nobile di Montepulciano", "Vino Nobile di Montepulciano", "IT"),
    # Spanish
    ("Rioja", "Rioja", "ES"),
    ("Sherry", "Jerez-Xérès-Sherry", "ES"),
    ("Jerez", "Jerez-Xérès-Sherry", "ES"),
    ("Cava", "Cava", "ES"),
    ("Priorat", "Priorat", "ES"),
    ("Priorato", "Priorat", "ES"),
    # Portuguese
    ("Port", "Porto", "PT"),
    ("Oporto", "Porto", "PT"),
    ("Douro", "Douro", "PT"),
    ("Madeira", "Madeira", "PT"),
    ("Vinho Verde", "Vinho Verde", "PT"),
    ("Dao", "Dão", "PT"),
    # German
    ("Mosel", "Mosel", "DE"),
    ("Moselle", "Mosel", "DE"),
    ("Rhine", "Rheingau", "DE"),
    ("Pfalz", "Pfalz", "DE"),
    ("Palatinate", "Pfalz", "DE"),
    # Austrian
    ("Wachau", "Wachau", "AT"),
    # Hungarian
    ("Tokaj", "Tokaji", "HU"),
    ("Tokay", "Tokaji", "HU"),
    # Greek
    ("Santorini", "\u03a3\u03b1\u03bd\u03c4\u03bf\u03c1\u03af\u03bd\u03b7", "GR"),
    ("Naoussa", "\u039d\u03ac\u03bf\u03c5\u03c3\u03b1", "GR"),
    ("Nemea", "\u039d\u03b5\u03bc\u03ad\u03b1", "GR"),
]


def main():
    parser = argparse.ArgumentParser(description="Seed appellation_aliases from primary sources")
    parser.add_argument("--dry-run", action="store_true", help="Preview without DB writes")
    args = parser.parse_args()

    print(f"\n{'=' * 60}")
    print("  APPELLATION ALIASES -- SEED FROM PRIMARY SOURCES")
    print(f"  {'(DRY RUN)' if args.dry_run else '(LIVE)'}")
    print(f"{'=' * 60}\n")

    sb = get_supabase()

    # Load all appellations with country info
    appellations = fetch_all("appellations", "id,name,designation_type,country_id")
    countries = fetch_all("countries", "id,name,iso_code")
    country_map = {c["id"]: c for c in countries}
    iso_to_country_id = {c["iso_code"]: c["id"] for c in countries}

    # Build appellation lookup by normalized name
    appellation_by_name: dict[str, dict] = {}
    appellation_by_norm: dict[str, dict] = {}
    for a in appellations:
        appellation_by_name[a["name"].lower()] = a
        appellation_by_norm[normalize(a["name"])] = a

    print(f"Loaded {len(appellations)} appellations from {len(countries)} countries\n")

    # Collect all aliases to insert
    aliases: list[dict] = []
    seen: set[str] = set()  # dedupe by alias_normalized

    def add_alias(appellation_id: str, alias: str, alias_type: str, source: str):
        norm = normalize(alias)
        # Skip if same as the canonical name
        appellation = next((a for a in appellations if a["id"] == appellation_id), None)
        if appellation and normalize(appellation["name"]) == norm:
            return
        # Skip duplicates
        if norm in seen:
            return
        seen.add(norm)
        aliases.append({
            "appellation_id": appellation_id,
            "alias": alias,
            "alias_normalized": norm,
            "alias_type": alias_type,
            "source": source,
        })

    # ====================================================================
    # SOURCE 1: INAO OpenDataSoft API -- French product variants
    # ====================================================================
    print("1. Fetching INAO product variants...")

    INAO_API = "https://public.opendatasoft.com/api/explore/v2.1/catalog/datasets/aires-et-produits-aocaop-et-igp/records"
    inao_products: list[dict] = []
    inao_offset = 0
    INAO_LIMIT = 100
    INAO_WHERE = "signe_fr LIKE 'AOC%' OR signe_fr LIKE 'IGP%'"

    with httpx.Client(timeout=30) as client:
        while True:
            params = {
                "select": "aire_geographique,produit,signe_fr,idproduit",
                "group_by": "aire_geographique,produit,signe_fr,idproduit",
                "where": INAO_WHERE,
                "limit": INAO_LIMIT,
                "offset": inao_offset,
            }
            resp = client.get(INAO_API, params=params)
            data = resp.json()
            results = data.get("results", [])
            if not results:
                break
            inao_products.extend(results)
            inao_offset += INAO_LIMIT
            if len(results) < INAO_LIMIT:
                break
            time.sleep(0.2)

    print(f"  Fetched {len(inao_products)} INAO wine product entries")

    # Match INAO products to our appellations
    inao_matched = 0
    inao_unmatched = 0
    inao_missed_set: set[str] = set()

    for p in inao_products:
        aire_norm = normalize(p["aire_geographique"])
        produit = p["produit"]
        produit_norm = normalize(produit)

        # Skip if produit is same as aire_geographique (no variant)
        if aire_norm == produit_norm:
            continue

        # Find matching appellation by aire_geographique
        appellation = appellation_by_norm.get(aire_norm)

        # If no direct match, try progressively shorter prefixes
        if not appellation:
            words = p["aire_geographique"].split(" ")
            for length in range(len(words) - 1, 0, -1):
                prefix = " ".join(words[:length])
                appellation = appellation_by_norm.get(normalize(prefix))
                if appellation:
                    break

        if appellation:
            # Determine alias type based on what's different
            alias_type = "synonym"
            diff = produit.replace(p["aire_geographique"], "").strip().lower()
            if diff in ("rouge", "blanc", "rosé", "clairet"):
                alias_type = "with_color"
            elif any(kw in diff for kw in (
                "vendanges tardives", "sélection de grains nobles",
                "vin jaune", "vin de paille", "mousseux", "primeur",
                "supérieur", "grand cru", "premier cru",
            )):
                alias_type = "with_designation"

            add_alias(appellation["id"], produit, alias_type, "inao-opendatasoft")
            inao_matched += 1
        else:
            inao_unmatched += 1
            inao_missed_set.add(p["aire_geographique"])

    print(f"  Matched: {inao_matched}, Unmatched: {inao_unmatched} ({len(inao_missed_set)} unique aires)")
    if inao_missed_set:
        sample = list(inao_missed_set)[:10]
        print(f"  Sample unmatched: {', '.join(sample)}")

    # ====================================================================
    # SOURCE 2: Tier 1 -- Mechanical color suffix generation
    # ====================================================================
    print("\n2. Generating Tier 1 color suffix aliases...")

    color_count = 0
    for a in appellations:
        country = country_map.get(a["country_id"])
        if not country or not country.get("iso_code"):
            continue
        suffixes = COLOR_SUFFIXES.get(country["iso_code"])
        if not suffixes:
            continue

        for suffix in suffixes:
            add_alias(a["id"], f"{a['name']} {suffix}", "with_color", "mechanical-color-suffix")
            color_count += 1

    pre_color = len(aliases) - color_count
    print(f"  Generated {color_count} color suffix candidates ({pre_color} already existed from INAO)")

    # ====================================================================
    # SOURCE 3: Tier 1 -- Accent-stripped variants
    # ====================================================================
    print("\n3. Generating accent-stripped aliases...")

    import unicodedata
    accent_count = 0
    for a in appellations:
        nfkd = unicodedata.normalize("NFD", a["name"])
        stripped = "".join(c for c in nfkd if unicodedata.category(c) != "Mn")
        if stripped != a["name"]:
            add_alias(a["id"], stripped, "synonym", "mechanical-accent-strip")
            accent_count += 1

    print(f"  Generated {accent_count} accent-stripped aliases")

    # ====================================================================
    # SOURCE 4: Tier 1 -- Designation type suffixes
    # ====================================================================
    print("\n4. Generating designation type suffix aliases...")

    designation_count = 0
    for a in appellations:
        dt = a.get("designation_type")
        if not dt:
            continue
        # Skip if the name already contains the designation
        if dt.lower() in a["name"].lower():
            continue
        add_alias(a["id"], f"{a['name']} {dt}", "with_designation", "mechanical-designation-suffix")
        designation_count += 1

    print(f"  Generated {designation_count} designation suffix aliases")

    # ====================================================================
    # SOURCE 5: Tier 2 -- Known translations
    # ====================================================================
    print("\n5. Adding known translation aliases...")

    translation_count = 0
    for english, local, iso in TRANSLATIONS:
        # Find the canonical appellation -- could be either the english or local name
        appellation = (
            appellation_by_name.get(local.lower())
            or appellation_by_name.get(english.lower())
            or appellation_by_norm.get(normalize(local))
            or appellation_by_norm.get(normalize(english))
        )
        if appellation:
            # Add the variant that ISN'T the canonical name
            if normalize(english) != normalize(appellation["name"]):
                add_alias(appellation["id"], english, "synonym", "known-translation")
                translation_count += 1
            if normalize(local) != normalize(appellation["name"]):
                add_alias(appellation["id"], local, "local_name", "known-translation")
                translation_count += 1

    print(f"  Added {translation_count} translation aliases")

    # ====================================================================
    # INSERT INTO DATABASE
    # ====================================================================
    print(f"\n{'-' * 60}")
    print(f"Total aliases to insert: {len(aliases)}")

    if args.dry_run:
        print("\n(DRY RUN -- not inserting)\n")
        # Show sample
        print("Sample aliases:")
        samples = [a for a in aliases if a["source"] == "inao-opendatasoft"][:10]
        for s in samples:
            app = next((a for a in appellations if a["id"] == s["appellation_id"]), None)
            print(f'  "{s["alias"]}" -> {app["name"] if app else "?"} [{s["alias_type"]}] ({s["source"]})')
        print("...")
        color_samples = [a for a in aliases if a["source"] == "mechanical-color-suffix"][:5]
        for s in color_samples:
            app = next((a for a in appellations if a["id"] == s["appellation_id"]), None)
            print(f'  "{s["alias"]}" -> {app["name"] if app else "?"} [{s["alias_type"]}] ({s["source"]})')

        # Stats by source
        by_src: dict[str, int] = {}
        for a in aliases:
            by_src[a["source"]] = by_src.get(a["source"], 0) + 1
        print("\nBy source:")
        for src, count in sorted(by_src.items(), key=lambda x: -x[1]):
            print(f"  {src}: {count}")
        return

    # Batch insert
    BATCH = 500
    inserted = 0
    for i in range(0, len(aliases), BATCH):
        batch = aliases[i:i + BATCH]
        try:
            sb.table("appellation_aliases").upsert(
                batch, on_conflict="alias_normalized"
            ).execute()
            inserted += len(batch)
        except Exception as e:
            print(f"  Batch error at {i}: {e}")
            # Try one by one for this batch
            for a in batch:
                try:
                    sb.table("appellation_aliases").upsert(
                        a, on_conflict="alias_normalized"
                    ).execute()
                    inserted += 1
                except Exception:
                    pass

    print(f"\nInserted {inserted} aliases")

    # Final stats
    try:
        result = sb.table("appellation_aliases").select("source").execute()
        by_src: dict[str, int] = {}
        for row in result.data or []:
            by_src[row["source"]] = by_src.get(row["source"], 0) + 1
        print("\nFinal counts by source:")
        for src, count in sorted(by_src.items(), key=lambda x: -x[1]):
            print(f"  {src}: {count}")
    except Exception:
        pass


if __name__ == "__main__":
    main()
